lib/LEDStatus/LEDStatus.cpp

```diff
@@ -0,0 +1,172 @@
+    _oneshotCountRemaining = count;
+    _oneshotCurrentlyOn = false;
+    // reset LED to off
+    if (_ledPin > 0)
+        digitalWrite(_ledPin, _pinState(LOW));
+    // restart timer
+    _timer = millis();
+}
+
+// call this function in your loop - it will return quickly after calculating if any changes need to 
+// be made to the pin to flash the LED
+void LEDStatus::LEDStatus::handle() {
+    // is a pin defined?
+    if (_ledPin == 0)
+        return;
+
+    // are we currently running a one-shot?
+    if (_oneshotCountRemaining > 0) {
+        if (_oneshotCurrentlyOn) {
+            if ((_timer + _oneshotOnMs) < millis()) {
+                if (_oneshotOffMs > 0)
+                    digitalWrite(_ledPin, _pinState(LOW));
+                _oneshotCurrentlyOn = false;
+                --_oneshotCountRemaining;
+                if (_oneshotCountRemaining == 0)
+                    _continuousCurrentlyOn = false;
+                _timer += _oneshotOnMs;
+            }
+        } else {
+            if ((_timer + _oneshotOffMs) < millis()) {
+                if (_oneshotOnMs > 0)
+                    digitalWrite(_ledPin, _pinState(HIGH));
+                _oneshotCurrentlyOn = true;
+                _timer += _oneshotOffMs;
+            }            
+        }
+    } else {
+        // operate using continuous
+        if (_continuousCurrentlyOn) {
+            if ((_timer + _continuousOnMs) < millis()) {
+                if (_continuousOffMs > 0)
+                    digitalWrite(_ledPin, _pinState(LOW));
+                _continuousCurrentlyOn = false;
+                _timer += _continuousOnMs;
+            }
+        } else {
+            if ((_timer + _continuousOffMs) < millis()) {
+                if (_continuousOnMs > 0)
+                    digitalWrite(_ledPin, _pinState(HIGH));
+                _continuousCurrentlyOn = true;
+                _timer += _continuousOffMs;
+            }
+        }
+    }
+}
+
+
+// private helper converts mode to on/off times in ms
+void LEDStatus::_modeToTime(LEDStatus::LEDMode mode, uint16_t& ledOffMs, uint16_t& ledOnMs) {
+    switch (mode) {
+        case LEDMode::Off:
+            ledOffMs = 1000;
+            ledOnMs = 0;
+            break;
+        case LEDMode::SlowToggle:
+            ledOffMs = 1000;
+            ledOnMs = 1000;
+            break;
+        case LEDMode::FastToggle:
+            ledOffMs = 100;
+            ledOnMs = 100;
+            break;
+        case LEDMode::SlowBlip:
+            ledOffMs = 1500;
+            ledOnMs = 50;
+            break;
+        case LEDMode::FastBlip:
+            ledOffMs = 333;
+            ledOnMs = 50;
+            break;
+        case LEDMode::On:
+            ledOffMs = 0;
+            ledOnMs = 1000;
+            break;
+        case LEDMode::Flicker:
+            ledOffMs = 50;
+            ledOnMs = 30;
+            break;
+        default:
+            Serial.printf("LEDStatus::_modeToTime: Uknown LED mode %d\n", mode);

```

Comment:
should this be `Unknown`?

Reply:
Think we should probably put the string in PROGMEM.

Reply:
Ah!  That's what F(xxx) does!  Will change.

Reply:
In case this saves you any time -- the syntax is slightly different with printf.  You'd use `printf_P` and the `PSTR` macro.